durable-storage/src/merkle_layer/node.rs

```diff
@@ -0,0 +1,194 @@
+// SPDX-FileCopyrightText: 2025 Trilitech <contact@trili.tech>
+//
+// SPDX-License-Identifier: MIT
+
+use std::fmt::Debug;
+
+use super::Key;
+
+/// A node which supports rebalancing operations
+pub(super) trait AvlNode: BinaryTreeNode + BinaryTreeNodeInvalidating {
+    /// Delete the node with a given key.
+    ///
+    /// If it does not exist, do nothing.
+    fn delete(_root: &mut Option<Box<Self>>, _key: &Key) {
+        todo!()
+    }
+
+    /// Return the height of the branch including this node.
+    #[cfg(test)] // Currently only used in a test
+    fn height(&self) -> u32;
+
+    /// Rebalance the node so that the difference in height between child branches is in the range
+    /// of -2..=2.
+    #[expect(dead_code, reason = "Not used")]
+    fn rebalance(_node: &mut Box<Self>) {
+        todo!()
+    }
+
+    /// Set the value of the node with a given key.
+    ///
+    /// If it does not exist, do nothing.
+    fn set(_root: &mut Option<Box<Self>>, _key: &Key, _data: Vec<u8>) {
+        todo!()
+    }
+}
+
+/// A node with basic immutable binary tree traversal.
+pub(super) trait BinaryTreeNode {
+    /// An immutable reference to the left branch.
+    fn left_ref(&self) -> &Option<Box<Self>>;
+
+    /// An immutable reference to the right branch.
+    fn right_ref(&self) -> &Option<Box<Self>>;
+}
+
+/// A node with basic mutable binary tree traversal and mutation.
+pub(super) trait BinaryTreeNodeInvalidating: NodeData + NodeDataInvalidating {
+    /// Delete the node with a given key.
+    ///
+    /// If it does not exist, do nothing.
+    fn delete(root: &mut Option<Box<Self>>, key: &Key);
+
+    /// A mutable reference to the left branch.
+    #[expect(dead_code, reason = "Not implemented")]
+    fn left_mut(&mut self) -> &mut Option<Box<Self>>;
+
+    /// A mutable reference to the right branch.
+    #[expect(dead_code, reason = "Not implemented")]
+    fn right_mut(&mut self) -> &mut Option<Box<Self>>;
+
+    /// Set the value of the node with a given key.
+    ///
+    /// If it does not exist, do nothing.
+    fn set(root: &mut Option<Box<Self>>, key: &Key, data: Vec<u8>);
+}
+
+/// A node that supports rebalancing and Merklisation
+#[derive(Clone, Default, Debug)]
+pub struct MavlNode {
+    key: Key,
+    data: Vec<u8>,
+    left: Option<Box<Self>>,
+    right: Option<Box<Self>>,
+
+    // A None hash is a hash that has not been set or has been dirtied
```

Comment:
Nit: Would like to have this in the tooltip.

```suggestion
    /// A [`None`] hash is a hash that has not been set or has been dirtied
```

Reply:
Not a problem [done](https://github.com/tezos/riscv-pvm/compare/41bd1267fc21076ce237a34a2ca200756abc4d24..d9f95ba4a91e7219e03649b5fcc1a6bd08ee1027#diff-afea9396726f50910e8482bf3de9c3c67054161c7619324e8a1f2bf0bd145e79R75)